File: src/main/java/com/dude/dms/ui/components/dialogs/EventDialog.java
package com.dude.dms.ui.components.dialogs;

import com.dude.dms.backend.data.DataEntity;
import com.dude.dms.ui.EntityCreateListener;
import com.dude.dms.ui.EntityDeleteListener;
import com.dude.dms.ui.EntityEditListener;
import com.vaadin.flow.component.dialog.Dialog;

public abstract class EventDialog<T extends DataEntity> extends Dialog {

    private EntityCreateListener<T> createListener;

    private EntityEditListener<T> editListener;

    private EntityDeleteListener<T> deleteListener;

    public void setCreateListener(EntityCreateListener<T> createListener) {
        this.createListener = createListener;
    }

    public void setDeleteListener(EntityDeleteListener<T> deleteListener) {
        this.deleteListener = deleteListener;
    }

    public void setEditListener(EntityEditListener<T> editListener) {
        this.editListener = editListener;
    }

    protected void triggerCreateEvent(T entity) {
        if (createListener != null) {
            createListener.onCreate(entity);
        }
    }

    protected void triggerDeleteEvent(T entity) {
        if (deleteListener != null) {
            deleteListener.onDelete(entity);
        }
    }

    protected void triggerEditEvent(T entity) {
        if (editListener != null) {
            editListener.onEdit(entity);
        }
    }
}

File: src/main/java/com/dude/dms/ui/components/dialogs/EventDialog.kt
package com.dude.dms.ui.components.dialogs

import com.dude.dms.brain.CreateEvent
import com.dude.dms.brain.DeleteEvent
import com.dude.dms.brain.EditEvent
import com.dude.dms.backend.data.DataEntity
import com.vaadin.flow.component.dialog.Dialog

abstract class EventDialog<T : DataEntity> : Dialog() {

    var createEvent: CreateEvent<T>? = null
    var editEvent: EditEvent<T>? = null
    var deleteEvent: DeleteEvent<T>? = null

    protected fun triggerCreateEvent(entity: T) {
        createEvent?.invoke(entity)
    }

    protected fun triggerDeleteEvent(entity: T) {
        deleteEvent?.invoke(entity)
    }

    protected fun triggerEditEvent(entity: T) {
        editEvent?.invoke(entity)
    }
}